### PR TITLE
Automatically choosing between algorithms for BC, BFS, SSSP, and TC on CPU

### PR DESCRIPTION
**Pull request by roshandathathri**

### CODE REVIEW

File: lonestar/analytics/cpu/sssp/SSSP.cpp [hunk 1]
```diff
@@ -80,8 +82,9 @@ static cll::opt<Algo>
                      clEnumVal(serDelta, "serDelta"),
                      clEnumVal(dijkstraTile, "dijkstraTile"),
                      clEnumVal(dijkstra, "dijkstra"), clEnumVal(topo, "topo"),
-                     clEnumVal(topoTile, "topoTile")),
-         cll::init(deltaTile));
+                     clEnumVal(topoTile, "topoTile"),
+                     clEnumVal(AutoAlgo, "choose between the algorithms automatically")),
```
Comment (l-hoang) [inline lonestar/analytics/cpu/sssp/SSSP.cpp:86]: nitpick: "among" algorithms, not between

File: lonestar/analytics/cpu/sssp/SSSP.cpp [hunk 2]
```diff
@@ -402,6 +405,15 @@ int main(int argc, char** argv) {
   galois::StatTimer execTime("Timer_0");
   execTime.start();
 
+  if (algo == AutoAlgo) {
+    if (isApproximateDegreeDistributionPowerLaw(graph)) {
+      algo = deltaStep;
+    } else {
+      algo = deltaStepBarrier;
+    }
+    galois::gInfo("Choosing ", algo, " algorithm");
```
Comment (l-hoang) [inline lonestar/analytics/cpu/sssp/SSSP.cpp:414]: same enum = int comment from BFS-dir-opt

**Comment by l-hoang** [comment]
Actually, I rescind my acceptance.

There seems to be a bunch of issues building as seen above by the CI: the RNG you're using doesn't work due to it being uninitialized.

File: lonestar/analytics/cpu/triangle-counting/Triangles.cpp [hunk 3]
```diff
@@ -462,7 +463,7 @@ void makeSortedGraph(Graph& graph) {
 }
 
 void readGraph(Graph& graph) {
-  if (relabel) {
+  if (relabel || isApproximateDegreeDistributionPowerLaw(graph)) {
```
[the remark this answers is not in the document]
Comment (roshandathathri) [reply lonestar/analytics/cpu/triangle-counting/Triangles.cpp:466]: Thanks!

File: lonestar/liblonestar/include/Lonestar/Search.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
+
+
+#pragma once
+#include <random>
+#include <vector>
+#include <algorithm>
+
+// Used to pick random non-zero degree starting points for search algorithms
```
[the remark this answers is not in the document]
Comment (roshandathathri) [reply lonestar/liblonestar/include/Lonestar/Search.h:8]: Will do. Thanks

File: lonestar/liblonestar/include/Lonestar/Search.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
+
```
[the remark this answers is not in the document]
Comment (roshandathathri) [reply lonestar/liblonestar/include/Lonestar/Search.h:1]: Thanks for the reminder

File: lonestar/analytics/cpu/bfs/bfsDirectionOpt.cpp [hunk 6]
```diff
@@ -387,6 +388,14 @@ void asyncAlgo(Graph& graph, GNode source, const P& pushWrap,
 
 template <bool CONCURRENT>
 void runAlgo(Graph& graph, const GNode& source, const uint32_t runID) {
+  if (algo == AutoAlgo) {
+    if (isApproximateDegreeDistributionPowerLaw(graph)) {
+      algo = SyncDO;
+    } else {
+      algo = Async;
+    }
+    galois::gInfo("Choosing ", algo, " algorithm");
```
[the remark this answers is not in the document]
Comment (roshandathathri) [reply lonestar/analytics/cpu/bfs/bfsDirectionOpt.cpp:397]: Yes, but that's fine for now.

File: lonestar/analytics/cpu/sssp/SSSP.cpp (hunk 1 above)
Comment (roshandathathri) [reply lonestar/analytics/cpu/sssp/SSSP.cpp:86]: Good point

**Comment by l-hoang** [comment]
One more comment:

are you reporting the chosen algorithm as a stat? it would help with parsing.

**Comment by roshandathathri** [comment]
> are you reporting the chosen algorithm as a stat? it would help with parsing.

Good point. But not right now. If it is done, it should be done for all applications so that it is consistent.

File: lonestar/liblonestar/include/Lonestar/Search.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
```
Comment (gurbinder533) [inline lonestar/liblonestar/include/Lonestar/Search.h:2]: Can't these routines be part of utils.h? search.h is slightly confusing.

Comment (roshandathathri) [reply lonestar/liblonestar/include/Lonestar/Search.h:2]: Ok. Will rename it as utils.h

Comment (l-hoang) [reply lonestar/liblonestar/include/Lonestar/Search.h:2]: There is already a utils.h somewhere is what Gurbinder is saying

Comment (roshandathathri) [reply lonestar/liblonestar/include/Lonestar/Search.h:2]: Not in liblonestar. I don't want to move this to libgalois or anywhere else.

File: lonestar/liblonestar/include/Lonestar/Utils.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    return source;
+  }
+};
+
+//! Used to determine if a graph has power-law degree distribution or not
+//! by sampling some of the vertices in the graph randomly
+//! This code has been copied from GAP benchmark suite
+//! (https://github.com/sbeamer/gapbs/blob/master/src/tc.cc WorthRelabelling())
+template <typename Graph>
+bool isApproximateDegreeDistributionPowerLaw(const Graph& graph) {
```
Comment (gurbinder533) [inline lonestar/liblonestar/include/Lonestar/Utils.h:36]: Maybe add a timer to measure the overhead?

**Pull request merged**